testsys: vmware-k8s for `cargo make test` (#2594)
bottlerocket-os/bottlerocket: TESTING.md

**Note:** For more information on publishing AMIs see [publishing](PUBLISHING.md).

### vmware-k8s

> You need to [build](BUILDING.md) Bottlerocket and a TUF repo to test VMware variant.

Reviewer: Need a line break.

```suggestion

You need to [build](BUILDING.md) Bottlerocket and a publicly accessible TUF repoistory to test VMware variants.
```
Can also hyperlink "TUF repo" to https://github.com/bottlerocket-os/bottlerocket/blob/develop/PUBLISHING.md#repo-location

Either `Infra.toml` or your environment need to be configured.
Make sure to set: GOVC_URL, GOVC_USERNAME, GOVC_PASSWORD, GOVC_DATACENTER, GOVC_DATASTORE, GOVC_NETWORK, GOVC_RESOURCE_POOL, GOVC_FOLDER.
If using `Infra.toml` the datacenter specified in `Test.toml` will be used, or the first one in the datacenters list of `Infra.toml` will be used if no datacenter is specified.
VMware testing also requires a `control-plane-endpoint` to be set in `Test.toml` for vmware cluster creation.
Change the commands below to the desired `vmware-k8s` variant:

```shell
cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  build

cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  repo
```

> Sync TUF repos with the vmware metadata and targets.

Reviewer: ```suggestion
Sync TUF repos containing the VMware variant's metadata and targets.
```

Reviewer 2: Is it clear what "sync TUF repos means"? What we really mean is, put the TUF repo in a place that the testsys resource agents can access via HTTP. And I guess this also needs to match the URLS in Infra.toml, right? Maybe say a little more here about what is happening for the uninitiated.

**Note:** For more information on publishing AMIs see [publishing](PUBLISHING.md).

### vmware-k8s

> First, a management cluster needs to be created using [`EKS Anywhere`](https://github.com/aws/eks-anywhere).

Reviewer: ```suggestion
First, an initial management cluster needs to be created using [`EKS Anywhere`](https://anywhere.eks.amazonaws.com/docs/getting-started/production-environment/vsphere-getstarted/#create-an-initial-cluster).
```

Maybe also mention setting a variable to point to the management cluster's kubeconfig file and we can pass that below in `cargo make test`

You need to [build](BUILDING.md) Bottlerocket and a TUF repo to test VMware variant.
Either `Infra.toml` or your environment need to be configured.
Make sure to set: GOVC_URL, GOVC_USERNAME, GOVC_PASSWORD, GOVC_DATACENTER, GOVC_DATASTORE, GOVC_NETWORK, GOVC_RESOURCE_POOL, GOVC_FOLDER.
If using `Infra.toml` the datacenter specified in `Test.toml` will be used, or the first one in the datacenters list of `Infra.toml` will be used if no datacenter is specified.

> VMware testing also requires a `control-plane-endpoint` to be set in `Test.toml` for vmware cluster creation.

Reviewer: ```suggestion
VMware testing also requires a `control-plane-endpoint` to be set in `Test.toml` for vSphere K8s cluster creation.
```

Change the commands below to the desired `vmware-k8s` variant:

```shell
cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  build

cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  repo

> ```

Reviewer: I recommend separating these two commands into two blocks and give a brief overview for both.

Sync TUF repos with the vmware metadata and targets.

```shell
cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  test \
  --mgmt-cluster-kubeconfig <PATH-TO-MGMT-CLUSTER-KUBECONFIG>
```

> 
> ```shell
> cargo make watch-test
> ```

Reviewer: ```suggestion

You can monitor the tests with:
```shell
cargo make watch-test
```
```

**Note:** For more information on publishing AMIs see [publishing](PUBLISHING.md).

### vmware-k8s

First, an initial management cluster needs to be created using [`EKS Anywhere`](https://anywhere.eks.amazonaws.com/docs/getting-started/production-environment/vsphere-getstarted/#create-an-initial-cluster).
You can then set `TESTSYS_MGMT_CLUSTER_KUBECONFIG` to the path to the management clusters kubeconfig.

> You need to [build](BUILDING.md) Bottlerocket and a publicly accessible [TUF repoistory](https://github.com/bottlerocket-os/bottlerocket/blob/develop/PUBLISHING.md#repo-location) to test VMware variants.

Reviewer: Ah sorry I typo'd my suggestion
```suggestion
You need to [build](BUILDING.md) Bottlerocket and a publicly accessible [TUF repository](https://github.com/bottlerocket-os/bottlerocket/blob/develop/PUBLISHING.md#repo-location) to test VMware variants.
```

Either `Infra.toml` or your environment need to be configured.
Make sure to set: GOVC_URL, GOVC_USERNAME, GOVC_PASSWORD, GOVC_DATACENTER, GOVC_DATASTORE, GOVC_NETWORK, GOVC_RESOURCE_POOL, GOVC_FOLDER.

> If using `Infra.toml` the datacenter specified in `Test.toml` will be used, or the first one in the datacenters list of `Infra.toml` will be used if no datacenter is specified.

Reviewer: For a bit more clarity
```suggestion
If `Infra.toml` is present, TestSys will use the data center specified in `Test.toml` first. If no data center is specified in `Test.toml`, then TestSys will use the first data center listed in `Infra.toml`.
```

VMware testing also requires a `control-plane-endpoint` to be set in `Test.toml` for vSphere K8s cluster creation.
Change the commands below to the desired `vmware-k8s` variant:

First, build the VMware variant you want to test.

```shell
cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  build
```

> Build the tuf repo containing the ova templates.

Reviewer: ```suggestion
Build the TUF repo containing the ova templates.
```

**Note:** For more information on publishing AMIs see [publishing](PUBLISHING.md).

### vmware-k8s

> First, an initial management cluster needs to be created using [`EKS Anywhere`](https://anywhere.eks.amazonaws.com/docs/getting-started/production-environment/vsphere-getstarted/#create-an-initial-cluster).

Reviewer 3: Is it a hard requirement to use EKS Anywhere to provision the management cluster? I could imagine using CAPV to provision a cluster instead. If it's just the recommended tool, we should soften the language here.

Author: As far as I know, an EKS Anywhere cluster created with the exact version as the one used in the agent must be used.

> First, an initial management cluster needs to be created using [`EKS Anywhere`](https://anywhere.eks.amazonaws.com/docs/getting-started/production-environment/vsphere-getstarted/#create-an-initial-cluster).

Reviewer 4: Does it have to be with `EKS Anywhere`? Is this the cluster to be tested or the cluster that runs the agents?

Regardless of the above, my understanding was that the OVAs can be used on-prem withouth having to use EKS Anywhere. Is testsys tightly coupled with EKS Anywhere?

Reviewer: This is the cluster that has the cluster API CRDs and manages other vSphere K8s cluster objects. This has to be created with EKS anywhere because there are special EKS-A CRDs and modifications to the upstream CAPI CRDs. Currently, to create vSphere K8s clusters with bottlerocket images, you have to use EKS-A since they provide the extra bootstrapping to get bottlerocket hosts to work as control plane nodes and etcd nodes. This is also likely how our vmware variant users are going to use the OVAs, to create vSphere K8s clusters via `eksctl anywhere`

You can then set `TESTSYS_MGMT_CLUSTER_KUBECONFIG` to the path to the management clusters kubeconfig.

> You need to [build](BUILDING.md) Bottlerocket and a publicly accessible [TUF repository](https://github.com/bottlerocket-os/bottlerocket/blob/develop/PUBLISHING.md#repo-location) to test VMware variants.

Reviewer 3: Is "publicly accessible" actually a requirement? It seems like an S3 bucket with the right policy to allow anonymous access through a VPC endpoint would work, for example.

Author: We can soften the language. In other places we do say publicly accessible, though.

Either `Infra.toml` or your environment need to be configured.

> Make sure to set: GOVC_URL, GOVC_USERNAME, GOVC_PASSWORD, GOVC_DATACENTER, GOVC_DATASTORE, GOVC_NETWORK, GOVC_RESOURCE_POOL, GOVC_FOLDER.

Reviewer 4: Are these environment variables the only ones that I need to set? In the documentation that you added there isn't any mention about these variables or if it is possible to set them up via `Infra.toml`.

If the `GOVC_` env variables can't be set via `Infra.toml`, make it clearer here, otherwise I suggest you add an snippet of code that shows how to use `Infra.toml` to set said variables.

What else does "your environment" covers? Are there environment variables that can be set for the data centers read from `Infra.toml`? If so, which ones?

Author: Those are the only ones needed to be set. The behavior is the same as `pubsys`, and documented in `Infra.toml.example`.

Reviewer 4: > The behavior is the same as pubsys, and documented in Infra.toml.example.

[This section] of the `Infra.toml.example` file covers the fields that will directly map to `GOVC_*`, that's fine. What is confusing from this wording is that it makes me think that I have to set `GOVC_*` as environment variables in my session in order to use testsys with VMware which (AFAIU) isn't actually true, since as you said I can use an existing `Infra.toml` configuration. So probably a bit of clarification will help, since saying `Either `Infra.toml` or your environment need to be configured.` 

```
Make sure you set X, Y, Z either via `Infra.toml` (link to the section that explains GOVC) or with environment variables in your session (if this is supported otherwise omit)
```

[This section]: https://github.com/bottlerocket-os/bottlerocket/blob/8d1ddbb5be6edb6f60ebf34e627403d3ec0108fb/tools/pubsys/Infra.toml.example#L73

Author: I have added some clarity.

> If `Infra.toml` is present, TestSys will use the data center specified in `Test.toml` first.

Reviewer 3: nit: elsewhere we refer to the command as `testsys` - we should be consistent

```suggestion
If `Infra.toml` is present, testsys will use the data center specified in `Test.toml` first.
```

> If `Infra.toml` is present, TestSys will use the data center specified in `Test.toml` first.
> If no data center is specified in `Test.toml`, then TestSys will use the first data center listed in `Infra.toml`.

Reviewer 4: The first line in the docs is a little bit "unusual", I read it as "if `Infra.toml` exists it will be ignored since `Test.toml` has higher precedence". You can rearrange the two sentences and avoid a few words to express the same precedence:

```suggestion
TestSys will use the data center speficied in `Test.toml` first.
If no data center is specified in `Test.toml`, Testsyswill use the first data center listed in `Infra.toml`
```

VMware testing also requires a `control-plane-endpoint` to be set in `Test.toml` for vSphere K8s cluster creation.
Change the commands below to the desired `vmware-k8s` variant:

First, build the VMware variant you want to test.

```shell
cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  build
```

Build the TUF repo containing the OVA templates.

```shell
cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  repo
```

> Sync TUF repos containing the VMware variant's metadata and targets. Make sure the TUF repos are accessible via HTTP and match the URLs in `Infra.toml`.

Reviewer 3: ```suggestion
Sync TUF repos containing the VMware variant's metadata and targets.
Make sure the TUF repos are accessible via unauthenticated HTTP or HTTPS and match the URLs in `Infra.toml`.
```

> Sync TUF repos containing the VMware variant's metadata and targets. Make sure the TUF repos are accessible via HTTP and match the URLs in `Infra.toml`.

Reviewer 4: nit: new line after `.`

Now, you can run the test.

```shell
cargo make \
  -e BUILDSYS_VARIANT="vmware-k8s-1.23" \
  -e BUILDSYS_ARCH="x86_64" \
  test \

>   --mgmt-cluster-kubeconfig $TESTSYS_MGMT_CLUSTER_KUBECONFIG

Reviewer 3: Can we enforce a convention where the management cluster is always the cluster where testsys is installed? It seems unfortunate to proliferate them, and makes it harder to reason about where the credentials come from to access the different clusters.

Author: I don't think we should require users to use an EKS Anywhere cluster for testsys if they _may_ want to test VMware variants.

Reviewer 3: I was thinking about the other direction - using `eksctl` to install the necessary EKS-A clusters onto an existing testsys cluster. That would simplify the setup process and reduce the number of "hops" between the developer's system and the actual provisioned nodes, which will help with troubeshooting if something goes wrong.

>   --mgmt-cluster-kubeconfig $TESTSYS_MGMT_CLUSTER_KUBECONFIG

Reviewer 4: nit:

```suggestion
  --mgmt-cluster-kubeconfig ${TESTSYS_MGMT_CLUSTER_KUBECONFIG}
```

**Note:** For more information on publishing AMIs see [publishing](PUBLISHING.md).

### vmware-k8s

First, an initial management cluster needs to be created using [`EKS Anywhere`](https://anywhere.eks.amazonaws.com/docs/getting-started/production-environment/vsphere-getstarted/#create-an-initial-cluster).
You can then set `TESTSYS_MGMT_CLUSTER_KUBECONFIG` to the path to the management clusters kubeconfig.
You need to [build](BUILDING.md) Bottlerocket and a publicly accessible [TUF repository](https://github.com/bottlerocket-os/bottlerocket/blob/develop/PUBLISHING.md#repo-location) to test VMware variants.
Either `Infra.toml` or your environment need to be configured.

> Make sure to set: GOVC_URL, GOVC_USERNAME, GOVC_PASSWORD, GOVC_DATACENTER, GOVC_DATASTORE, GOVC_NETWORK, GOVC_RESOURCE_POOL, GOVC_FOLDER.

Reviewer 3: Nit: I'd put this in a list or code block for readability.

```suggestion
Make sure to set the following environment variables:

- GOVC_URL
- GOVC_USERNAME
- GOVC_PASSWORD
- GOVC_DATACENTER
- GOVC_DATASTORE
- GOVC_NETWORK
- GOVC_RESOURCE_POOL
- GOVC_FOLDER
```

If `Infra.toml` is present, testsys will use the data center specified in `Test.toml` first.

> Testsys will use the data center speficied in `Test.toml` first.

Reviewer 5: s/speficied/specified/
